sdk/storage/azure-storage-quickquery/src/main/java/com/azure/storage/quickquery/BlobQuickQueryInputStream.java

```diff
@@ -0,0 +1,207 @@
+            return -1;
+        } else {
+            throw new IOException("Error parsing blob quick query stream. No end record was present. This may indicate "
+                + "that not all data was returned.");
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
+        return this.read(new byte[1], 0, 1);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (this.parsedStream != null) {
+            this.parsedStream.close();
+        }
+        if (this.buffer != null) {
+            buffer.close();
+        }
+        super.close();
+    }
+
+    /**
+     * Validates that all parameters are non-null. Throws IOException if any of them are.
+     */
+    private boolean checkParametersNotNull(String record, Object... data) throws IOException {
+        for (Object o : data) {
+            if (o == null) {
+                throw new IOException("Failed to parse " + record + " record from blob query response stream.");
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Parses a generic avro record to known data types.
+     * @param record {@link GenericRecord}
+     * @throws IOException on failed parsing.
+     */
+    private void parseRecord(GenericRecord record) throws IOException {
+        String schemaName = record.getSchema().getName();
+
+        switch (schemaName) {
+            case "resultData": {
+                Object data = record.get("data");
+
+                if (checkParametersNotNull("result data", data)) {
+                    this.buffer = new ByteArrayInputStream(((ByteBuffer) data).array());
+                }
+                break;
+            } case "end": {
+                if (this.progressReceiver != null) {
+                    Object total = record.get("totalBytes");
+
+                    if (checkParametersNotNull("end", total)) {
+                        this.progressReceiver.reportProgress((Long) total);
+                    }
+                }
+                this.endRecordSeen = true;
+                break;
+            } case "progress": {
+                if (this.progressReceiver != null) {
+                    Object scanned = record.get("bytesScanned");
+                    Object total = record.get("totalBytes");
+
+                    if (checkParametersNotNull("progress", scanned, total)) {
+                        this.progressReceiver.reportProgress((Long) scanned);
+                    }
+                }
+                break;
+            } case "error": {
+                Object fatal = record.get("fatal");
+                Object name = record.get("name");
+                Object description = record.get("description");
+                Object position = record.get("position");
+
+                if (checkParametersNotNull("error", fatal, name, description, position)) {
+                    BlobQuickQueryError error = new BlobQuickQueryError((Boolean) fatal, name.toString(),
```

Comment:
This is quite confusing. The parameter client takes is named `nonFatalErrorReceiver` but we're passing both fatals and non-fatals there.